### ISSUE TITLE
Smartcard login results in IRP failure

### ISSUE THREAD

**Issue by hedgeberg** [FreeRDP/FreeRDP#4709]
Hey guys,

Been debugging this for a bit now to try and pin this down. During an attempt to log in while running 2.0.0-rc2 (or any other release), I get to the log-in screen on the target, but the smartcard log in option does not appear on the target, despite the reader visibly becoming active on the client machine. 

OS is Arch Linux. Here's a uname output in case that helps:

```
hedge@blade ~/D/r/FreeRDP> uname -a
Linux blade 4.16.13-2-ARCH #1 SMP PREEMPT Fri Jun 1 18:46:11 UTC 2018 x86_64 GNU/Linux
```

I'm running freerdp with the following:

```
> ./client/X11/xfreerdp -sec-nla /smartcard:"Yubico Yubikey NEO CCID 00 00" /v:192.168.1.20 
```

Which produces the following response in stdout: 

```
[19:00:43:136] [18014:18015] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
[19:00:43:136] [18014:18015] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
[19:00:43:136] [18014:18015] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[19:00:43:158] [18014:18015] [ERROR][com.winpr.timezone] - Unable to find a match for unix timezone: US/Eastern
[19:00:43:212] [18014:18015] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[19:00:43:212] [18014:18015] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[19:00:43:241] [18014:18015] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[19:00:43:242] [18014:18024] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard [Yubico Yubikey NEO CCID 00 00] (static)
[19:00:43:421] [18014:18024] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
[19:00:43:484] [18014:18038] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardStatusW (0x000900CC), status: SCARD_E_INSUFFICIENT_BUFFER (0x80100008)
```

At this point, I get a session with the login screen on the target, but no option to log in w/ smartcard, and after closing the session I get:

```
[19:00:47:513] [18014:18015] [INFO][com.freerdp.client.x11] - Closed from X11
[19:00:47:513] [18014:18030] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[19:00:47:514] [18014:18027] [ERROR][com.freerdp.channels.rdpdr.client] - pVirtualChannelWriteEx failed with CHANNEL_RC_BAD_CHANNEL_HANDLE [00000007]
[19:00:47:514] [18014:18027] [ERROR][com.freerdp.channels.smartcard.client] - smartcard_complete_irp failed with error 7!
```

In the interest of completeness, here's lsusb:

```
hedge@blade ~/D/r/FreeRDP> lsusb
***SNIP***
Bus 001 Device 012: ID 1050:0112 Yubico.com Yubikey NEO(-N) CCID
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

And here's the output from pcsc_scan:

```
hedge@blade ~/D/r/FreeRDP> pcsc_scan 
Using reader plug'n play mechanism
Scanning present readers...
0: Yubico Yubikey NEO CCID 00 00
 
Sat Jun 16 19:10:21 2018
 Reader 0: Yubico Yubikey NEO CCID 00 00
  Card state: Card inserted, 
  ATR: 3B FC 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 4E 45 4F 72 33 E1

ATR: 3B FC 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 4E 45 4F 72 33 E1
+ TS = 3B --> Direct Convention
+ T0 = FC, Y(1): 1111, K: 12 (historical bytes)
  TA(1) = 13 --> Fi=372, Di=4, 93 cycles/ETU
    43010 bits/s at 4 MHz, fMax for Fi = 5 MHz => 53763 bits/s
  TB(1) = 00 --> VPP is not electrically connected
  TC(1) = 00 --> Extra guard time: 0
  TD(1) = 81 --> Y(i+1) = 1000, Protocol T = 1 
-----
  TD(2) = 31 --> Y(i+1) = 0011, Protocol T = 1 
-----
  TA(3) = FE --> IFSC: 254
  TB(3) = 15 --> Block Waiting Integer: 1 - Character Waiting Integer: 5
+ Historical bytes: 59 75 62 69 6B 65 79 4E 45 4F 72 33
  Category indicator byte: 59 (proprietary format)
+ TCK = E1 (correct checksum)

Possibly identified card (using /usr/share/pcsc/smartcard_list.txt):
3B FC 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 4E 45 4F 72 33 E1
	YubiKey NEO (PKI)
	http://www.yubico.com/
```

Log with /log-level=trace and stdout + stderr redirected to log.txt is here:

[log.txt](https://github.com/FreeRDP/FreeRDP/files/2108569/log.txt)

Let me know what other debugging information I can get and/or where to poke. For what its worth, I'm fully aware Arch makes this kind of thing weird and that the issue is potentially in an outside library, and I'm willing to experiment to fix this. Just want to not need to reboot into windows every time I need to do admin stuff on my NAS.

**Comment by hedgeberg** [comment]
Should add that smartcard login works perfectly using the Windows RDP client, and I have this card provisioned etc. Neither rdesktop nor freerdp play nice with my setup though, and they dont play nice with any other smartcard that I have either. Thanks!

**Comment by akallabeth** [comment]
@hedgeberg smartcard login with nla is not implemented.

**Comment by akallabeth** [comment]
@hedgeberg Currently smartcard is only enabled for redirection (so everything that does not authenticate before you see a screen)
to enable smartcard authentication you need to implement the sections in `[MS-CSSP]` (`2.2.1.2.2 TSSmartCardCreds` and others).
Don't know how hard it is to translate the tokens from `pcsc-lite` to something `ms` compatible.

**Comment by hedgeberg** [comment]
Wouldn't it be possible to authenticate using smart card redirection after establishing securities? I.e. establish the connection without auths, bring up the login screen, then simulate a smart card plug-n-play event? At least as a workaround? 

Forgive me if this is a dumb idea, I know remarkably little about RDP, up until now I'm mostly just a user. I'm more of an embedded developer than an app/OS level dev, so I'm sure I'm going to end up asking more than a few dumb questions. 

As is, -sec-nla with no /u or /d flags (see example in initial post, but without the /smartcard flag) brings up a login screen from which I can log in using password login without any prior request, but even with /smartcard enabled windows doesn't seem to detect it. 

I'll take a look at the MS spec mentioned in the morning, and I'll ping a windows internals guy I know to see if I can get any background on how windows handles PIV tokens etc.

**Comment by hedgeberg** [comment]
(side note, is there anywhere I can go to ask questions other than GitHub? I spent some time hanging out in #freerdp but it was pretty dead)

**Comment by informatimago** [comment]
Note that I have a patch with smartcard logon via RDP working nicely on this branch: https://github.com/informatimago/FreeRDP/tree/smartcard-logon-test 
I'm waiting on #4701 to be merged to make the PR.
Have a look at: https://github.com/informatimago/FreeRDP/wiki/smartcard-logon
I'm currently working on making smartcard logon work with NLA too.

**Comment by dodo040** [comment]
My PR #3946 is still pending. Lot of work have been done since its opening. You can find a more recent branch here (last rebase from FreeRDP master dated to 17 august 2017) : https://github.com/dodo040/FreeRDP-smartcardlogon/commits/master).
But a rebase is needed to be up-to-date since significant changes on smartcard have been done in the master at the end of year 2017. I'll try to allocate time to work on this in the weeks (or months) to come.

**Comment by hedgeberg** [comment]
@informatimago I'll give it a shot, but in your patch is the certificate transmitted in plaintext? does the connection remain plaintext/no security after initial response? Considering that PIV by nature is hardware-backed, this isn't a huge risk, but it does definitely leave me feeling a little queasy about the idea of remoting in to my home network with this.

@dodo040 got a summary on hand of what the changes in your branch vs. where it was rebased off of are? I can look at rebasing your fork later this week, if you're comfortable with that.

**Comment by informatimago** [comment]
@hedgeberg indeed, currently only smartcard thru RDP is implemented. You may still use it on an encrypted VPN (ssh tunnel, etc).

**Comment by hedgeberg** [comment]
@akallabeth forgive the nagging, when I run 

```
xfreerdp -sec-nla /v:192.168.1.20
```
I get a login window without requiring username, password, or domain, and I can then log on from the Windows login prompt. However, with ```/smartcard``` added, I still don't see a smartcard sign-in option appear. My assumption was that smartcard traffic would be forwarded over the tunnel, meaning that, to windows, the smart card should appear as though i was plugged in and signing into the machine locally, no? Is smartcard forwarding completely disabled under /sec-nla? if so, this is extremely confusing, considering that xfreerdp queries a significant amount of information from the smartcard according to my pcscd logs.

**Comment by akallabeth** [comment]
@hedgeberg don't exactly know what has to be enabled on the windows server to allow smartcard redirection on the login screen, the tunnel needs a server side component enabled.
`nla` does authentication before establishing the connection, so you don't get a login screen.

**Comment by martinl** [comment]
Hi,

I have been following the smartcard login support related developments in FreeRDP for a while and would like to see it implemented in near future. So I set up a $100 bounty on Bountysource to encourage getting smartcard login with NLA support implemented in FreeRDP (and ideally, included in the 2.0.0 release version). 

Feel free to join in as backer or as developer resolving the issue: https://www.bountysource.com/issues/59769663-smartcard-login-results-in-irp-failure

Bounty will be paid to developer who creates smartcard login with NLA support PR after it is merged to FreeRDP master branch and verified to be working. 

I will use Debian 10 or macOS Mojave as FreeRDP client, Windows Server 2016 as RDP server and Yubikey NEO (PIV+opensc) as smartcard/usb token for verifying the implemented functionality.

If I underestand correctly, most of the work would be related to cleaning up https://github.com/FreeRDP/FreeRDP/pull/5195. Also, there is some related information in https://github.com/FreeRDP/FreeRDP/issues/3239

PS. not sure how/why FreeRDP repo issues are related to Remmina on Bountysource, maybe you can fix it as repo owners

with best regards,
Martin

**Comment by akallabeth** [comment]
#7620 implements NLA with smartcard, so closing older issues referring to that.
